daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -661,9 +661,13 @@ private[lf] object Speedy {
     @inline
     private[speedy] final def kontDepth(): Int = kontStack.size()
 
-    private[speedy] def asUpdateMachine(location: String)(f: UpdateMachine => Control): Control
+    private[speedy] def asUpdateMachine(location: String)(
+        f: UpdateMachine => Control[Question.Update]
+    ): Control[Q]
```

Comment:
We just cannot. `Q` is the type parameter of the `Machine[Q]` in case of `PureMachine` `Q = Nothing` and in case  of `ScenarioMachine` `Q  = Question.Scenario`. 

This method actually implement a runtime check that the machine is actually an `UpdateMachine` , if yes `Q = Question.Update` otherwise the machine crash. 

Basically, the change of this PR it just guarantee the machine outputs only the question it is concern about (UpdateMachine outputs Update questions, ScenarioMachine outputs Scenario questions, PureMachine output no question). It is not concern about the concern about avoiding preexisting runtime check. 

The only to avoid this kind of run time check would be two have different set of continuations, builtins, expressions for each type of machine, this is definitively too much effort, and does not prevent the machine if some other invariant is broken.

Reply:
Thanks for that explanation. I'd guessed that there was a good reason for not doing things like this, but it wasn't immediately obvious to myself.